Migrate standalone controllers in pages to component directives.

YashJipkate opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
This PR migrates the standalone controllers in the following pages to component directives. 

- about page
- Creator dashboard
- Donate page
- Email dashboard pages
- Topic landing pages
- Error page
- Learner Dashboard
- Moderator Page
- Notification Dashboard
- Preferences
- Practice Session
- Splash Page

<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->

## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [ ] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR has an appropriate "CHANGELOG: ..." label (If you are unsure of which label to add, ask the reviewers for guidance).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR addresses the points mentioned in the codeowner checks for the files/folders changed. (See the [codeowner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).)
- [x] The PR is **assigned** to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer and don't tick this checkbox.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue. Do not only request the review but also add the reviewer as an assignee.

YashJipkate committed:
Rename and files about page

YashJipkate committed:
Convert to component directive about page

YashJipkate committed:
Merge branch 'change-name-branch' into comp-dir-page

YashJipkate committed:
About Page

Fix lint

Fix lint

YashJipkate committed:
Merge branch 'develop' into comp-dir-page

YashJipkate committed:
Modify controller

YashJipkate committed:
renaming and files creator dashboard

YashJipkate committed:
Merge branch 'change-name-branch' into comp-dir-page

YashJipkate committed:
Wrap up creator dashboard

YashJipkate committed:
Merge develop

YashJipkate committed:
Donate page

YashJipkate committed:
Fix backend tests and about -> about-page, creator-dashboard -> creator-dashboard-page

YashJipkate committed:
Email Dashboard

YashJipkate committed:
Email dashboard Result

YashJipkate committed:
Lint

seanlip reviewed (changes requested):
@YashJipkate, some files in this PR are showing up as add/delete. Could you please look into that, or make smaller PRs?

YashJipkate commented:
Yes the controller files for the three pages are showing as add/delete and the rest are good. That is the reason I migrated only three standalone controllers of small pages.

Should I make even smaller migrations?

seanlip commented:
Yes. The goal of breaking down is to ensure that nothing shows up as an add/delete when it should really show up as a move.

YashJipkate commented:
Actually I tried doing that and it still showed as add/delete. While on my local it was shown as rename/modify.

seanlip commented:
Even when you migrate just a single page? Can you link to that PR?

YashJipkate committed:
Modify

YashJipkate committed:
Wrap up

YashJipkate committed:
Wrap up topic landing

YashJipkate committed:
Lint

YashJipkate commented:
Hi @seanlip I have migrated a fresh page [here](https://github.com/YashJipkate/oppia/pull/29) and it shows add/delete while I specifically used git mv.

seanlip commented:
I think that's because you're trying to change a controller into a directive and move things around at the same time. That's why git thinks it's a different file.

Change the controller into the directive first. Then do a separate PR for the move.

YashJipkate committed:
Change .directive -> .controller for about page

YashJipkate committed:
Change .directive -> .controller for creator dashboard and donate

YashJipkate commented:
@seanlip The files are now in modify-mode although the names are *.controller.ts. 
You can take a look now!
Thanks!

YashJipkate committed:
Merge branch 'develop' into comp-dir-page

YashJipkate committed:
Topic landing page and creator dashboard karma fix and merge develop

seanlip reviewed (approved):
LGTM, thanks!

If you would like to make some follow-up commits to this PR which do the move, I can review those too.

YashJipkate committed:
Fix backend topic editor

YashJipkate committed:
Merge branch 'more-compdir' into comp-dir-page

YashJipkate committed:
.directive -> .controller email dashboard pages

YashJipkate committed:
Error page full

YashJipkate committed:
Stewards landing page

YashJipkate committed:
Leaner dashboard

YashJipkate committed:
Fix backend

YashJipkate committed:
Fix lint

YashJipkate committed:
Merge develop

YashJipkate committed:
Correct path

YashJipkate committed:
Moderator page

YashJipkate committed:
Notifications Dashboard page

YashJipkate committed:
Preferences page

YashJipkate committed:
Revert "Revert "Practice session page""

This reverts commit bfa7107e91540d12c9b2b1d106da3dc4967d1223.

YashJipkate committed:
Practice session page

YashJipkate committed:
Revert "Practice session page"

This reverts commit 26713bd7cfeb87ce876985fe8e51fe0d72811832.

YashJipkate committed:
Splash page

YashJipkate commented:
@seanlip I have done some more pages. PTAL!

seanlip commented on core/controllers/custom_landing_pages_test.py:
```diff
@@ -57,30 +57,3 @@ def test_invalid_topic_landing_page_leads_to_404(self):
         response.mustcontain('<topic-landing-page></topic-landing-page>')
-
-
-class StewardsLandingPageTest(test_utils.GenericTestBase):
```
Why are you just deleting this? That's not right.

YashJipkate replied:
This function looks for strings in the code. But since they are not loaded anymore, as directives are not loaded in backend tests. So it would just have `<stewards-landing-page></srewards-landing-age>`. Should I test for this then?

seanlip replied:
Yes.

YashJipkate replied:
I have added them. PTAL!

seanlip reviewed (changes requested):
OK, I slightly regret saying what I did, because reviewing the following commits was quite messy. Let's fix the one comment I have and get this PR in without further additions.

Also, please make sure to manually test all the pages; we just found a bug in QA that was due to replacing scope with ctrl. Make sure to actually take a look at the pages visually.

YashJipkate committed:
RE-added the backend test

seanlip reviewed (approved):
Thanks! LGTM.

vinitamurthi reviewed (commented):
Overall looks good, have you tested the practice sessions page? We had some previous refactors break practice sessions and since we cannot have e2e tests for this right now it has to be manually tested.

set ENABLE_NEW_STRUCTURES_PLAYERS = true in constants.js
Then add a new topic(lets say TestTopic) and assign a skill to it, create 1-2 questions in that topic and then you can go to 
localhost:<port>/practice_session/TestTopic

YashJipkate commented:
HI, @vinitamurthi Yes I checked it and it was working fine. If you can, you can also test this just to be sure since you have a lot better understanding of this structure. Although I will do some more testing.
Thanks!

seanlip commented:
Please note that Travis checks are failing.

YashJipkate committed:
Fix e2e in preferences

YashJipkate commented:
@seanlip The tests failed due to the use of `$parent` in preferences-page.directive.html. I replaced the `ctrl` with `$scope` to solve the issue. Though I am a bit concerned about the use of `$parent`. There is an issue in the styleguide about this [here](https://github.com/johnpapa/angular-styleguide/issues/457). 

@vinitamurthi Thanks for pointing out on the practice-sessions page. There seems to be something off with how the conversation-skin.directive is not able to get the context of question player and keeps throwing console errors. It might be because the conversation-skin.directive is not a component directive. But that is just a hypothesis that might be terribly wrong. I have reverted the changes to practice session as of now.

EDIT: I found that the same issue is on develop and not on my branch. This is the log that shows up.
![Screenshot from 2019-06-14 15-02-01](https://user-images.githubusercontent.com/34203227/59499747-a598f380-8eb5-11e9-92af-dd750081323e.png)
It appears when I complete all the questions and press 'SUBMIT'. @vinitamurthi @seanlip Could you PTAL if this is a valid bug. If yes then this PR has nothing to do with it. Also note that this was a fresh build with dist/ and local_compiled_js/ deleted.

seanlip commented:
For the question player, I have to defer to @vinitamurthi. 

But you should definitely remove the use of $parent altogether, that's part of the point of this refactor to make things into directives. Pass the necessary info via directive attributes instead. Make sure to carefully test manually that all the functionality still works fine.

vinitamurthi commented:
Hi @YashJipkate that error is fine, that is because the work of the result player is yet to be checked in so the submit button cannot go anywhere right now. I just wanted to verify that the question player does get loaded and you can answer questions, thanks for checking that!

seanlip commented:
@YashJipkate I don't think you addressed my latest commit. We should be using ctrl and not $parent.

YashJipkate committed:
removed

YashJipkate commented:
@seanlip @kevinlee12 I have removed `$parent` and the tests now pass too ([in my local](https://travis-ci.com/YashJipkate/oppia/builds/115673509)). Could you PTAL while it passes here.

seanlip reviewed (approved):
Thanks @YashJipkate, LGTM. You manually tested all the Preferences page functionality right?

YashJipkate commented:
Yes @seanlip I have tested the functionality. Though if someone could point out a mistake, it would be a great help!

vojtechjelinek commented on core/templates/dev/head/pages/about-page/about-page.scripts.ts:
```diff
@@ -0,0 +1,24 @@
+require('base_components/BaseContentDirective.ts');
+require(
+  'components/common-layout-directives/common-elements/' +
+  'background-banner.directive.ts');
```
This block of requires should be in `pages/about-page/about-page.controller.ts` because the appropriate tags for directives are in `pages/about-page/about-page.directive.html`, do the same for the other files.

vojtechjelinek commented on core/templates/dev/head/pages/about-page/about-page.controller.ts:
```diff
@@ -12,91 +12,99 @@
-require(
-  'components/common-layout-directives/common-elements/' +
-  'background-banner.directive.ts');
-
```
This should have been kept in this file. Same for the other pages.

YashJipkate replied:
Done. The other pages dont have a controller file so they are in the scripts file.

vojtechjelinek commented:
It seems to me that most of the files still have wrong renames, like `error-page.controller.ts → email-dashboard-page.scripts.ts`

YashJipkate committed:
Modify imports

YashJipkate commented:
> It seems to me that most of the files still have wrong renames, like `error-page.controller.ts → email-dashboard-page.scripts.ts`

It is actually super weird. Idk how this can possibly happen. Especially if they are done both in separate commits.

EDIT: Fixed it.

vojtechjelinek commented on core/templates/dev/head/pages/about-page/about-page.controller.ts:
```diff
@@ -12,91 +12,102 @@
+ * @fileoverview Controllers for the about page.
+ */
+
+require('domain/utilities/UrlInterpolationService.ts');
```
This `require()` should be after the `background-banner.directive.ts` `require()` with one empty line between these two. More is explained in https://github.com/oppia/oppia/wiki/Coding-style-guide#webpack

YashJipkate replied:
Done!

vinitamurthi commented:
Hi @YashJipkate looks good to me, just want to confirm that practice sessions page has been manually tested right?

YashJipkate committed:
Modify imports

YashJipkate commented:
Yes @vinitamurthi Apart from the console error everything as you said works fine!

vojtechjelinek commented on core/templates/dev/head/pages/error-pages/error-page.scripts.ts:
```diff
@@ -0,0 +1,20 @@
+ */
+
+require('pages/OppiaFooterDirective.ts');
+require('pages/error-pages/error-page.directive.ts');
```
Alphabetical order, please.

vojtechjelinek replied:
Alphabetical order, please.

YashJipkate committed:
Modify imports

vojtechjelinek reviewed (approved):
LGTM!

YashJipkate committed:
Modify imports

YashJipkate committed:
.directive -> .controller error-page

kevinlee12 reviewed (approved):
lgtm!

seanlip commented:
@YashJipkate Please note merge conflicts.

kevinlee12 commented:
@YashJipkate, mind the merge conflicts!

seanlip commented:
@nithusha21 PTAL

YashJipkate committed:
Merge conflicts

YashJipkate committed:
Merge conflicts

YashJipkate commented:
@seanlip @kevinlee12 The merge conflicts are resolved and the tests pass ([on my local](https://travis-ci.com/YashJipkate/oppia/builds/115698709)). PTAL!

seanlip commented:
OK, thanks! Though I think it's @nithusha21 you want a review from, not us.

YashJipkate commented:
OK @seanlip, @nithusha21 PTAL! Thanks!

YashJipkate commented:
@nithusha21 Ping!

nithusha21 commented on core/templates/dev/head/pages/practice-session-page/practice-session-page.directive.html:
```diff
@@ -0,0 +1,33 @@
+  }
+
+  @media(max-width: 925px) {
+    .oppia-practice_session-containerr {
```
Is this a typo? It should be a single r at the end right?

YashJipkate replied:
I don't know, I'll have to check since this was in the original file. I did not modify this. Most probably this is a typo.

nithusha21 commented on core/templates/dev/head/pages/preferences-page/preferences-page.controller.spec.ts:
```diff
@@ -30,7 +31,8 @@ describe('Preferences Controller', function() {
     beforeEach(angular.mock.inject(function(
-        $controller, $http, _$httpBackend_, $rootScope) {
+        _$componentController_, $http, _$httpBackend_, $rootScope) {
+      $componentController = _$componentController_;
```
What does the _ at both ends mean?

YashJipkate replied:
I saw this here - > https://docs.angularjs.org/guide/component#unit-testing-component-controllers

nithusha21 commented on core/templates/dev/head/pages/splash-page/splash-page.directive.html:
```diff
@@ -0,0 +1,165 @@
+<div>
```
Do you need this external div?

YashJipkate replied:
This was present in the original file so I decided not to tamper with it lest it should break the UI.

nithusha21 reviewed (approved):
Not very familiar with this migration process, but just for my understanding.... Are we converting all ng-controller blocks to use a directive instead? I am ok to merge this PR once you confirm that you have visited each of these pages being migrated, and have manually tested that it's not broken. Approving as a codeowner.

Please give a confirmation of the testing before merging. In particular, at lease make sure to manually test all critical user journeys that are affected by the PR. I will link the doc with all the paths listed in  a follow up comment.

nithusha21 commented:
Here is the doc with critical user journeys: https://docs.google.com/document/d/1T3HyMU8cMvXY1tyzs801Zgf5oSxLqaHICUH_YZJa4JM/edit?usp=drivesdk

YashJipkate committed:
Merge develop

YashJipkate committed:
Fix merge

YashJipkate commented:
Hi @nithusha21 I have tested all the critical user paths affected by this PR and no errors are thrown. PTAL!

YashJipkate commented:
Hi @nithusha21 The tests now pass! PTAL!

nithusha21 commented:
Thanks! Also, for the 2 nits mentioned above, can you file a separate PR? I don't want to block this PR on that, but that needs to be fixed. Can you do that ASAP (and it would be easier to test as you change only 1 page).

YashJipkate commented:
Hi @nithusha21 I have filed #6954. PTAL!

nithusha21 commented:
Hi @YashJipkate, after this was merged, the build on travis is broken https://travis-ci.org/oppia/oppia/builds/546720893. Can you fix it ASAP to unblock all contributors (prioritize this).

YashJipkate commented:
Hi @nithusha21 The build fail is due to Travis and not the PR. If you look at the log it has some wierd error on third party script which is untouched by this PR. If anyone is facing failing builds then they can restart the particular failing job (note that this job also differs in various build failures caused due to this this error, further indicating problem with Travis itself).

Thanks!

nithusha21 commented:
I'm not too sure what the weird error is... Could you by any chance add the log itself over here? Also, yeah I think we can wait to see if restarting helped. But it has been failing consistently across all builds, and hence might be an issue with our code.

YashJipkate commented:
Here is the log - > https://travis-ci.org/oppia/oppia/jobs/546868171#L1351
